src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/Gen2GcCallback.cs

```diff
@@ -0,0 +1,119 @@
+            _weakTargetObj = GCHandle.Alloc(targetObj, GCHandleType.Weak);
+        }
+
+        /// <summary>
+        /// Schedule 'callback' to be called in the next GC.  If the callback returns true it is
+        /// rescheduled for the next Gen 2 GC.  Otherwise the callbacks stop.
+        /// </summary>
+        public static void Register(Func<bool> callback)
+        {
+            // Create a unreachable object that remembers the callback function and target object.
+            _ = new Gen2GcCallback(callback);
+        }
+
+        /// <summary>
+        /// Schedule 'callback' to be called in the next GC.  If the callback returns true it is
+        /// rescheduled for the next Gen 2 GC.  Otherwise the callbacks stop.
+        ///
+        /// NOTE: This callback will be kept alive until either the callback function returns false,
+        /// or the target object dies.
+        /// </summary>
+        public static void Register(Func<object, bool> callback, object targetObj)
+        {
+            // Create a unreachable object that remembers the callback function and target object.
+            _ = new Gen2GcCallback(callback, targetObj);
+        }
+
+        ~Gen2GcCallback()
+        {
+            if (_weakTargetObj.IsAllocated)
+            {
+                // Check to see if the target object is still alive.
+                object? targetObj = _weakTargetObj.Target;
+                if (targetObj == null)
+                {
+                    // The target object is dead, so this callback object is no longer needed.
+                    _weakTargetObj.Free();
+                    return;
+                }
+
+                // Execute the callback method.
+                try
+                {
+                    Debug.Assert(_callback1 != null);
+                    if (!_callback1(targetObj))
+                    {
+                        // If the callback returns false, this callback object is no longer needed.
+                        return;
+                    }
+                }
+                catch
+                {
+                    // Ensure that we still get a chance to resurrect this object, even if the callback throws an exception.
+#if DEBUG
+                    // Except in DEBUG, as we really shouldn't be hitting any exceptions here.
+                    throw;
+#endif
+                }
+            }
+            else
+            {
+                // Execute the callback method.
+                try
+                {
+                    Debug.Assert(_callback0 != null);
+                    if (!_callback0())
+                    {
+                        // If the callback returns false, this callback object is no longer needed.
+                        return;
+                    }
+                }
+                catch
+                {
+                    // Ensure that we still get a chance to resurrect this object, even if the callback throws an exception.
```

Comment:
er... not sure how much i love this.  like... if throws... it's in some sort of bad state.  keeping smething alive in a bad state seems like a literal case of creating zombies :D

Reply:
➡️ Keeping this one as-is to minimize deviations from dotnet/runtime